### PR TITLE
Operator overload in reverse mode

### PR DESCRIPTION
**Pull request by PhrygianGates**
Add support for operator overload in reverse mode

### CODE REVIEW

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 1]
```diff
@@ -1993,6 +2016,18 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
           }
         }
       }
+
+      if (opCode == UnaryOperatorKind::UO_Deref &&
+          m_Mode == DiffMode::reverse_mode_forward_pass) {
```
Comment (parth-07) [inline lib/Differentiator/ReverseModeVisitor.cpp:2021]: Why do we have reverse mode forward pass specific code in ReverseModeVisitor? This defeats the whole purpose of having a separate class for ReverseModeVisitor.

**Review by parth-07 — CHANGES_REQUESTED** [review]
We need more tests. 
1) Member operator overloads -- operator overloads that are defined inside the class.
2) Operator overload that does not return a reference type.
3) Unary operator overload.

**Comment by PhrygianGates** [comment]
> We need more tests.
> 
> 1. Member operator overloads -- operator overloads that are defined inside the class.
> 2. Operator overload that does not return a reference type.
> 3. Unary operator overload.

For 1, our tests already cover the case.
For 2, because we have this prior restriction in the existing code we can not support the case
```cpp
      if (!FD->getReturnType()->isVoidType()) {
        assert((pullback && !FD->getReturnType()->isVoidType()) &&
               "Call to function returning non-void type with no dfdx() is not "
               "supported!");
      }
```
 For 3, I have added tests for unary op.

**Comment by parth-07** [comment]
> For 2, because we have this prior restriction in the existing code base, we can not support the case yet.

Why does the restriction that you mentioned prevent non-reference return types? Also, why do you think it only affects non-reference return type operator overloads, but not non-reference return type ordinary member functions?

File: benchmark/VectorModeComparison.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
+#include "benchmark/benchmark.h"
+
+#include "clad/Differentiator/Differentiator.h"
+
+#include "BenchmarkedFunctions.h"
+
+// Benchmark forward mode for weighted sum.
+static void BM_ForwardModeWeightedSum(benchmark::State& state) {
```
Comment (parth-07) [inline benchmark/VectorModeComparison.cpp:8]: These benchmarks shouldn't be part of this pull request. Can you please rebase the pull request properly?

**Review by vgvassilev — APPROVED** [review]
LGTM!

**Pull request merged**